storage/diskmetricstore.go

```diff
@@ -246,7 +246,12 @@ func (dms *DiskMetricStore) persist() error {
 	}
 	inProgressFileName := f.Name()
 	e := gob.NewEncoder(f)
-	if err := e.Encode(dms.metricGroups); err != nil {
+
+	if err := func() error {
+		dms.lock.RLock()
+		defer dms.lock.RUnlock()
+		return e.Encode(dms.metricGroups)
+	}(); err != nil {
```

Comment:
It's fine to write it more straight forward:

```go
dms.lock.RLock()
err = e.Encode(dms.metricGroups)
dms.lock.RUnlock()
if err != nil {
// ...
```

err already exists in the scope.